dev/com.ibm.ws.repository.resolver/test/com/ibm/ws/repository/resolver/ResolutionTests.java

```diff
@@ -1912,11 +1914,179 @@ public void testCircleNotToRoot() throws Exception {
+     *
+     * Feautre I is private
+     *
+     * Dependency map
+     * Feature A -> Feature X -> Feature I 1.0
+     * Feature B -> Feature Y -> Feature I 1.1 tolerate 1.0
+     *
+     * Expected: resolution succeeds because although Feature B does not tolerate Feature I 1.0, Feature I is private.
+     */
+    @Test
+    public void testSetTransitivePrivate() throws RepositoryException {
+        // initialize all the features
+        EsaResourceWritable featureA = WritableResourceFactory.createEsa(null);
+        featureA.setProvideFeature("com.example.featureA-1.0");
+        featureA.addRequireFeatureWithTolerates("com.example.featureX-1.0", Collections.<String> emptyList());
+        featureA.setVisibility(Visibility.PUBLIC);
+        repoFeatures.add(featureA);
+
+        EsaResourceWritable featureB = WritableResourceFactory.createEsa(null);
+        featureB.setProvideFeature("com.example.featureB-1.0");
+        featureB.addRequireFeatureWithTolerates("com.example.featureY-1.0", Collections.<String> emptyList());
+        featureB.setVisibility(Visibility.PUBLIC);
+        repoFeatures.add(featureB);
+
+        EsaResourceWritable featureX = WritableResourceFactory.createEsa(null);
+        featureX.setProvideFeature("com.example.featureX-1.0");
+        featureX.addRequireFeatureWithTolerates("com.example.featureI-1.0", Collections.<String> emptyList());
+        featureX.setVisibility(Visibility.PUBLIC);
+        repoFeatures.add(featureX);
+
+        EsaResourceWritable featureY = WritableResourceFactory.createEsa(null);
+        featureY.setProvideFeature("com.example.featureY-1.0");
+        featureY.addRequireFeatureWithTolerates("com.example.featureI-1.1", Arrays.asList("1.0"));
+        featureY.setVisibility(Visibility.PUBLIC);
+        repoFeatures.add(featureY);
+
+        EsaResourceWritable featureI10 = WritableResourceFactory.createEsa(null);
+        featureI10.setProvideFeature("com.example.featureI-1.0");
+        featureI10.setVisibility(Visibility.PRIVATE);
+        featureI10.setSingleton("true");
+        repoFeatures.add(featureI10);
+
+        EsaResourceWritable featureI11 = WritableResourceFactory.createEsa(null);
+        featureI11.setProvideFeature("com.example.featureI-1.1");
+        featureI11.setVisibility(Visibility.PRIVATE);
+        featureI11.setSingleton("true");
+        repoFeatures.add(featureI11);
+
+        // build the repository and resolve the features
+        RepositoryResolver resolver = createResolver();
+        Collection<List<RepositoryResource>> resolved = resolver.resolveAsSet(Arrays.asList(featureA.getProvideFeature(), featureB.getProvideFeature()));
+        assertThat(resolved, containsInAnyOrder(
+                                                Matchers.<RepositoryResource> contains(featureI10, featureX, featureA),
+                                                Matchers.<RepositoryResource> contains(featureI10, featureY, featureB)));
+
+    }
+
```

Comment:
Is it worth also having a test where 
     * Dependency map
     * Feature A -> Feature X -> Feature I 1.1
     * Feature B -> Feature Y -> Feature I 1.0 tolerate 1.1

I.e. B uses an older level but tolerates a newer level, does this make any difference to the logic being tested?

Reply:
Yeah, I can add this